crates/ascot-os/src/mac.rs

```diff
@@ -0,0 +1,327 @@
+                    "Failed for prefix {prefix:02X?}"
+                );
+            }
+        }
+
+        #[test]
+        fn test_is_not_virtual_mac_vendor() {
+            assert!(!is_virtual_mac_vendor([0x00, 0x1A, 0x2B, 0, 0, 0]));
+            assert!(!is_virtual_mac_vendor([0xFF, 0xFF, 0xFF, 0, 0, 0]));
+        }
+    }
+}
+
+#[cfg(target_os = "windows")]
+#[allow(unsafe_code)]
+mod windows {
+    use std::{mem, ptr};
+
+    use tracing::warn;
+
+    use windows_sys::Win32::Foundation::ERROR_SUCCESS;
+    use windows_sys::Win32::NetworkManagement::IpHelper::{
+        GetAdaptersAddresses, GetIfEntry2, IF_TYPE_ETHERNET_CSMACD, IF_TYPE_IEEE80211,
+        IP_ADAPTER_ADDRESSES_LH, MIB_IF_ROW2,
+    };
+    use windows_sys::Win32::NetworkManagement::Ndis::{
+        IfOperStatusUp, NdisPhysicalMedium802_3 as NDIS_PHYSICAL_MEDIUM802_3,
+        NdisPhysicalMediumNative802_11 as NDIS_PHYSICAL_MEDIUM_NATIVE802_11,
+    };
+    use windows_sys::Win32::Networking::WinSock::AF_UNSPEC;
+
+    // Extracts MAC address from a network interface row if it's up and has a valid 6-byte address.
+    fn extract_mac_from_row(row: &MIB_IF_ROW2) -> Option<[u8; 6]> {
+        if row.OperStatus == IfOperStatusUp && row.PhysicalAddressLength == 6 {
+            let mut mac = [0u8; 6];
+            mac.copy_from_slice(&row.PhysicalAddress[..6]);
+            Some(mac)
+        } else {
+            None
+        }
+    }
+
+    // Traverses the adapter linked list and extracts the first matching Wi-Fi and Ethernet MAC addresses.
+    fn process_adapter(
+        adapter: *mut IP_ADAPTER_ADDRESSES_LH,
+    ) -> (Option<[u8; 6]>, Option<[u8; 6]>) {
+        let mut wifi = None;
+        let mut ethernet = None;
+
+        let mut current = adapter;
+        while !current.is_null() {
+            // SAFETY: `current` is a valid pointer to an IP_ADAPTER_ADDRESSES_LH structure.
+            // The list is well-formed and terminated with a null pointer.
+            let addr = unsafe { &*current };
+
+            // SAFETY: `row` is zero-initialized and safe to pass to GetIfEntry2,
+            // which will write valid data into this structure.
+            let mut row: MIB_IF_ROW2 = unsafe { mem::zeroed() };
+            row.InterfaceLuid = addr.Luid;
+
+            // Populate row with interface information based on adapter LUID.
+            //
+            // SAFETY: GetIfEntry2 is called with a valid pointer to `row`.
+            // Return value 0 indicates success.
+            if unsafe { GetIfEntry2(&mut row) } == 0 {
+                if let Some(mac) = extract_mac_from_row(&row) {
+                    match (row.Type, row.PhysicalMediumType) {
+                        // Verify interface type and physical medium.
+                        (IF_TYPE_IEEE80211, NDIS_PHYSICAL_MEDIUM_NATIVE802_11) => wifi = Some(mac),
+                        (IF_TYPE_ETHERNET_CSMACD, NDIS_PHYSICAL_MEDIUM802_3) => {
+                            ethernet = Some(mac)
```

Comment:
Perhaps I'm wrong, but if there are **more** `wifi` and `ethernet` interfaces, we just retrieve the last ones. In our case we are not interested in having more addresses, but it would be better to add a comment explaining this concept